### PR TITLE
[FSDP] Construct FQN in _full_post_state_dict_hook

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #81253

Differential Revision: [D37079444](https://our.internmc.facebook.com/intern/diff/D37079444/)

### CODE REVIEW

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -1921,33 +1922,59 @@ def _full_post_state_dict_hook(
...
         if not state_dict:
             return state_dict
 
+        if self._fsdp_wrapped_module.no_params:
+            return state_dict
+
+        # This can happen when rank0_only is enabled and self.rank != 0.
+        if self._fsdp_wrapped_module.orig_flat_param[0] is None:
+            assert self._fsdp_wrapped_module.flat_param is not None, (
+                "When no_params is False, one of flat_param and orig_flat_param "
```
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:1938]: Is this a developer or user facing issue? If user can hit this issue, might be good to mention explicitly what went wrong and raise a Runtime/ValueError.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 2]
```diff
@@ -1921,33 +1922,59 @@ def _full_post_state_dict_hook(
...
+
+        # Loop only the parameters saved in self._fsdp_wrapped_module to avoid
+        # processing buffers.
+        for module_name, _, param_name in (
+            self._fsdp_wrapped_module.orig_flat_param[0].param_info
+        ):
+            module_name = module_name.replace(f"{FPW_MODULE}.", "")
+            module_name = module_name.replace(f"{FPW_MODULE}", "")
+            if module_name:
+                module_name = f"{module_name}."
```
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:1954]: is there a reason we need to postfix a "." here and `param_info` does not provide it like this? Seems quite easy to forget.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 3]
```diff
@@ -1921,33 +1922,59 @@ def _full_post_state_dict_hook(
...
-                # the original buffer's GPU memory and require reallocating
-                # that memory later; this only affects the state dict's buffer
-                # variable and leaves the original buffer's GPU memory intact
-                if offload_to_cpu and state_dict[key].device != cpu_device:
-                    state_dict[key] = state_dict[key].to(cpu_device)
-                continue
+
             # Clone non-ignored parameters before exiting the
             # `_summon_full_params()` context
+            key = fqn
```
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:1972]: just use `clean_key` here?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 4]
```diff
@@ -1962,7 +1989,21 @@ def _full_post_state_dict_hook(
...
+        # that memory later; this only affects the state dict's buffer
+        # variable and leaves the original buffer's GPU memory intact
+        if offload_to_cpu:
+            for clean_key in self._buffer_names:
+                # This is a hack to support activation checkpoint.
+                clean_key = clean_key.replace(
+                    f"{checkpoint_wrapper._CHECKPOINT_PREFIX}.", ""
+                )
+                fqn = f"{prefix}{clean_key}"
+                fqn = fqn.replace("mod.", "")
```
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:2004]: this should not be needed now?

**Review by rohan-varma — APPROVED** [review]
Some minor comments and "mod" prefixing should not be needed now. Thanks!

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 2 above)
Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:1954]: I prefer that `param_info` keeps the `module_name` just as the module name without the `.`.

IMO, it makes more logical sense that `module_name` is just the module's name. The `.` is only relevant when prefixing for a parameter name, which may not be the only reason we access `module_name`.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 1 above)
Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:1938]: It should be internal.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 5]
```diff
@@ -1921,33 +1922,59 @@ def _full_post_state_dict_hook(
...
             return state_dict
 
+        if self._fsdp_wrapped_module.no_params:
+            return state_dict
+
+        # This can happen when rank0_only is enabled and self.rank != 0.
+        if self._fsdp_wrapped_module.orig_flat_param[0] is None:
+            assert self._fsdp_wrapped_module.flat_param is not None, (
+                "When no_params is False, one of flat_param and orig_flat_param "
+                "should has value."
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:1939]: nit:
```suggestion
                "should have value."
```

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 1 above)
Comment (fegin) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:1938]: Should be an assert here as this should not happen (anymore). If it happens, it should be an assert.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 2 above)
Comment (fegin) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:1954]: Will keep this statement. `param_info` should keeps the `module_name` as the actual module name.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 3 above)
Comment (fegin) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:1972]: Using `clean_key` is not correct as `clean_key` may be stripped out. Remove this key and directly use `fqn`.

**Comment by fegin** [comment]
#pytorchbot merge

**Pull request closed without merging**